src/api/app/controllers/webui/watch_items_controller.rb

```diff
@@ -0,0 +1,58 @@
+class Webui::WatchItemsController < ApplicationController
+  before_action :check_item_and_user, only: [:create]
+  before_action :check_item_to_destroy, only: [:destroy]
+
+  # TODO: Add a JSON response that includes item name, url,...
+  # or whatever other information that we need in the JS watchlist
+
+  def create
+    watch_item = WatchItem.new(item: @item, user: @user)
+    respond_to do |format|
+      if watch_item.save
+        format.json { render json: @user.watch_items }
+      else
+        format.json { render json: { error: watch_item.errors.to_json, status: :unprocessable_entity } }
+      end
+    end
+  end
+
+  def destroy
+    respond_to do |format|
+      if @item_to_destroy.destroy
+        format.json { render json: @item_to_destroy.user.watch_items }
+      else
+        format.json { render json: @item_to_destroy.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  private
+
+  def check_item_and_user
+    @item = case params[:item_type]
+            when 'project'
+              Project.find(params[:item_id])
+            when 'package'
+              Package.find(params[:item_id])
+            when 'request'
+              BsRequest.find(params[:item_id])
+            end
+    @user = if params[:user_id].present?
```

Comment:
Yes, I guess we can omit it if we are only permitting the current user to modify his watchlist

Reply:
Sure only you can change your watchlist.

As general feedback: it's never a good idea to lump `before_actions` that gather data together. Separate concerns will give you structure and you're users the right feedback in case of failure.